### PR TITLE
adding ip check for annotatePod in ipamd

### PR DESCRIPTION
**Pull request by jerryhe1999**
**What type of PR is this?**
bug

**Which issue does this PR fix**:
#2281 

**What does this PR do / Why do we need it**:
This PR adds a check login in annotatePod() of ipamd.go to make sure when deleting the Pod IP, the IP is the one we are releasing.

**Testing done on this change**:
New unit test added in ipamd_test.go. Both unit tests and integration tests passed.

**Automation added to e2e**:
N/A

**Will this PR introduce any new dependencies?**:
No

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:
No

**Does this change require updates to the CNI daemonset config files to work?**:
No

**Does this PR introduce any user-facing change?**:
No

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/ipamd/ipamd_test.go [hunk 1]
```diff
@@ -2344,3 +2344,97 @@ func TestIsConfigValid(t *testing.T) {
...
+
+	updatedPod, err = mockContext.GetPod(pod.Name, pod.Namespace)
+	assert.NoError(t, err)
+	assert.NotNil(t, updatedPod.Annotations["ip-address"])
+	assert.Equal(t, "10.0.0.4", updatedPod.Annotations["ip-address"])
+
+	// Test DEL operation
+	// shouldn't delete
+	err = mockContext.AnnotatePod(pod.Name, pod.Namespace, "ip-address", "", "10.0.0.1")
+	assert.Error(t, err)
```
Comment (jdn5126) [inline pkg/ipamd/ipamd_test.go:2425]: Let's check that the value of `err` here is as expected

File: pkg/ipamd/rpc_handler.go [hunk 2]
```diff
@@ -179,7 +179,7 @@ func (s *server) AddNetwork(ctx context.Context, in *rpc.AddNetworkRequest) (*rp
 	}
 
 	if s.ipamContext.enablePodIPAnnotation {
-		s.ipamContext.AnnotatePod(in.K8S_POD_NAME, in.K8S_POD_NAMESPACE, vpccniPodIPKey, ipv4Addr)
```
Comment (jdn5126) [inline pkg/ipamd/rpc_handler.go:182]: nit: let's add a comment here that on ADD, we pass empty string as there is no IP being released

File: pkg/ipamd/rpc_handler.go [hunk 3]
```diff
@@ -273,7 +273,7 @@ func (s *server) DelNetwork(ctx context.Context, in *rpc.DelNetworkRequest) (*rp
 	}
 
 	if s.ipamContext.enablePodIPAnnotation {
-		s.ipamContext.AnnotatePod(in.K8S_POD_NAME, in.K8S_POD_NAMESPACE, vpccniPodIPKey, "")
+		s.ipamContext.AnnotatePod(in.K8S_POD_NAME, in.K8S_POD_NAMESPACE, vpccniPodIPKey, "", ip)
```
Comment (jdn5126) [inline pkg/ipamd/rpc_handler.go:276]: nit: let's add a comment that on DEL, we pass IP being released

**Review by jdn5126 — CHANGES_REQUESTED** [review]
Changes are looking good, mostly nits. Before submitting, can you make sure `make check-format` passes? You can run `go fmt` on the files you modified to get the formatting fixed.

**Comment by jerryhe1999** [comment]
@jdn5126 Have updated a new commit, could you have look on it? Thank you!

File: pkg/ipamd/ipamd.go [hunk 4]
```diff
@@ -2069,12 +2069,31 @@ func (c *IPAMContext) AnnotatePod(podName, podNamespace, key, val string) error
 		if newPod.Annotations == nil {
 			newPod.Annotations = make(map[string]string)
 		}
-		newPod.Annotations[key] = val
+		// On CNI ADD, always set new annotation
```
Comment (jdn5126) [inline pkg/ipamd/ipamd.go:2072]: nit: this comment should go below the next line

File: pkg/ipamd/ipamd.go [hunk 5]
```diff
@@ -2069,12 +2069,31 @@ func (c *IPAMContext) AnnotatePod(podName, podNamespace, key, val string) error
...
+		oldVal, ok := newPod.Annotations[key]
+		if newVal != "" {
+			//If the new pod annotation is as the same as the old one, return immediately without extra cost
```
Comment (jdn5126) [inline pkg/ipamd/ipamd.go:2075]: nit: space after comment, and maybe rephrase it as "Skip patch operation if new value is the same as existing value"

File: pkg/ipamd/ipamd_test.go [hunk 6]
```diff
@@ -2344,3 +2344,84 @@ func TestIsConfigValid(t *testing.T) {
...
+	assert.Equal(t, ipTwo, updatedPod.Annotations["ip-address"])
+
+	//Test that add operation always overwrites value for existing pod
+	err = mockContext.AnnotatePod(pod.Name, pod.Namespace, "ip-address", ipThree, ipTwo)
+	assert.NoError(t, err)
+
+	//Test that delete operation will not overwrite if IP being released does not match existing value
+	err = mockContext.AnnotatePod(pod.Name, pod.Namespace, "ip-address", "", ipOne)
+	assert.Error(t, err)
+	assert.Equal(t, fmt.Errorf("Released IP 10.0.0.1 does not match existing annotation. Not patching pod."), err)
```
Comment (jdn5126) [inline pkg/ipamd/ipamd_test.go:2409]: nit: use `ipOne` here to simplify check

File: pkg/ipamd/ipamd_test.go [hunk 7]
```diff
@@ -2344,3 +2344,84 @@ func TestIsConfigValid(t *testing.T) {
...
+		enableIPv6:      false,
+	}
+
+	mockContext.rawK8SClient.Create(ctx, &pod)
+
+	ipOne := "10.0.0.1"
+	ipTwo := "10.0.0.2"
+	ipThree := "10.0.0.3"
+
+	//Test basic add operation for new pod
```
Comment (jdn5126) [inline pkg/ipamd/ipamd_test.go:2379]: nit: let's add a space after comment. I am surprised this didn't get done by `go fmt`, but maybe you have an older go installation

**Review by jdn5126 — CHANGES_REQUESTED** [review]
A few last nits, then this looks good to me

**Review by jdn5126 — CHANGES_REQUESTED** [review]

**Review by jdn5126 — APPROVED** [review]
